sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -0,0 +1,367 @@
+          allChildStagesMaterialized = results.forall(_.allChildStagesMaterialized),
+          newStages = results.flatMap(_.newStages))
+      }
+  }
+
+  private def newQueryStage(e: Exchange): QueryStageExec = {
+    val optimizedPlan = AdaptiveSparkPlanExec.applyPhysicalRules(e.child, queryStageOptimizerRules)
+    val queryStage = e match {
+      case s: ShuffleExchangeExec =>
+        ShuffleQueryStageExec(currentStageId, s.copy(child = optimizedPlan))
+      case b: BroadcastExchangeExec =>
+        BroadcastQueryStageExec(currentStageId, b.copy(child = optimizedPlan))
+    }
+    currentStageId += 1
+    queryStage
+  }
+
+  private def reuseQueryStage(s: QueryStageExec, output: Seq[Attribute]): QueryStageExec = {
+    val queryStage = ReusedQueryStageExec(currentStageId, s, output)
+    currentStageId += 1
+    queryStage
+  }
+
+  /**
+   * Update the logical plan after new query stages have been created and the physical plan has
+   * been updated with the newly created stages.
+   * 1. If the new query stage can be mapped to an integral logical sub-tree, replace the
+   *    corresponding logical sub-tree with a leaf node [[LogicalQueryStage]] referencing the new
+   *    query stage. For example:
+   *        Join                   SMJ                      SMJ
+   *      /     \                /    \                   /    \
+   *    r1      r2    =>    Xchg1     Xchg2    =>    Stage1     Stage2
+   *                          |        |
+   *                          r1       r2
+   *    The updated plan node will be:
+   *                               Join
+   *                             /     \
+   *    LogicalQueryStage1(Stage1)     LogicalQueryStage2(Stage2)
+   *
+   * 2. Otherwise (which means the new query stage can only be mapped to part of a logical
+   *    sub-tree), replace the corresponding logical sub-tree with a leaf node
+   *    [[LogicalQueryStage]] referencing to the top physical node into which this logical node is
+   *    transformed during physical planning. For example:
+   *     Agg           HashAgg          HashAgg
+   *      |               |                |
+   *    child    =>     Xchg      =>     Stage1
+   *                      |
+   *                   HashAgg
+   *                      |
+   *                    child
+   *    The updated plan node will be:
+   *    LogicalQueryStage(HashAgg - Stage1)
+   */
+  private def updateLogicalPlan(newStages: Seq[(Exchange, QueryStageExec)]): Unit = {
+    newStages.foreach { n =>
+      val (oldNode: SparkPlan, newNode: SparkPlan) = n
+      val logicalNodeOpt = oldNode.logicalLink.orElse(oldNode.collectFirst {
+        case p if p.logicalLink.isDefined => p.logicalLink.get
+      })
+      assert(logicalNodeOpt.isDefined)
+      val logicalNode = logicalNodeOpt.get
+      val physicalNode = currentPhysicalPlan.collectFirst {
+        case p if p.eq(newNode) || p.logicalLink.exists(logicalNode.eq) => p
```

Comment:
It's the most complicated part here:
Normally we only need to replace an Exchange with a logical "Stage", simple straightforward logic. But when it comes to several physical nodes generated from a single logical node and there's an Exchange is in the middle of these nodes, we need to apply a little "trick" here, by replacing the top node (e.g., the final agg) with a logical "Stage" which wraps the whole sub-tree starting from the top node and of course containing the exchange stage.

The code comment for this method explains this logic. Alternatively, we could have logical aggregate that represents "final" and "local" aggregate in order not to go this route. Let's have a follow-up on that though.

Reply:
Since this is hacky, how about we explicitly match the physical aggregate (hash and sort) here? Except for aggregate, we should always look for condition `if p.eq(newNode)`